example_graceful_shutdown_test.go

```diff
@@ -0,0 +1,149 @@
+	ctx := context.Background()
+
+	dbPool, err := pgxpool.NewWithConfig(ctx, riverinternaltest.DatabaseConfig("river_testdb_example"))
+	if err != nil {
+		panic(err)
+	}
+	defer dbPool.Close()
+
+	// Required for the purpose of this test, but not necessary in real usage.
+	if err := riverinternaltest.TruncateRiverTables(ctx, dbPool); err != nil {
+		panic(err)
+	}
+
+	jobStarted := make(chan struct{})
+
+	workers := river.NewWorkers()
+	river.AddWorker(workers, &WaitsForCancelOnlyWorker{jobStarted: jobStarted})
+
+	riverClient, err := river.NewClient(riverpgxv5.New(dbPool), &river.Config{
+		Logger: slog.New(&slogutil.SlogMessageOnlyHandler{Level: slog.LevelWarn}),
+		Queues: map[string]river.QueueConfig{
+			river.DefaultQueue: {MaxWorkers: 100},
+		},
+		Workers: workers,
+	})
+	if err != nil {
+		panic(err)
+	}
+
+	_, err = riverClient.Insert(ctx, WaitsForCancelOnlyArgs{}, nil)
+	if err != nil {
+		panic(err)
+	}
+
+	if err := riverClient.Start(ctx); err != nil {
+		panic(err)
+	}
+
+	riverClientStopped := make(chan struct{})
+
+	sigterm := make(chan os.Signal, 1)
+	signal.Notify(sigterm, syscall.SIGTERM)
+
+	// This is meant to be a realistic-looking stop goroutine that might go in a
+	// real program. It waits for SIGTERM and when received, tries to stop
+	// gracefully by allowing a chance for jobs to finish. But if that isn't
+	// working, a second SIGTERM will tell it to terminate with prejudice and
+	// it'll issue a hard stop that cancels the context of all active jobs.
+	go func() {
+		defer close(riverClientStopped)
+
+		<-sigterm
+		fmt.Printf("Received SIGTERM; initiating soft stop (try to wait for jobs to finish)\n")
+
+		softStopSucceeded := make(chan struct{})
+		go func() {
+			if err := riverClient.Stop(ctx); err != nil {
+				if !errors.Is(err, context.Canceled) {
+					panic(err)
+				}
+			}
+			close(softStopSucceeded)
+		}()
+
+		select {
+		case <-sigterm:
+			fmt.Printf("Received SIGTERM again; initiating hard stop (cancel everything)\n")
+			if err := riverClient.StopAndCancel(ctx); err != nil {
+				panic(err)
+			}
+		case <-softStopSucceeded:
+			// Will never be reached in this example.
+		}
+	}()
+
+	// Make sure our job starts being worked before doing anything else.
+	<-jobStarted
+
+	// Cheat a little by sending a SIGTERM manually for the purpose of this
+	// example (normally this will be sent by user or supervisory process). The
+	// first SIGTERM tries a soft stop in which jobs are givcen a chance to
+	// finish up.
+	sigterm <- syscall.SIGTERM
+
+	// The soft stop will never work in this example because our job only
+	// respects context cancellation, but wait a short amount of time to give it
+	// a chance. After it elapses, send another SIGTERM to initiate a hard stop.
+	select {
+	case <-riverClientStopped:
+		// Will never be reached in this example because our job will only ever
+		// finish on context cancellation.
+		fmt.Printf("Soft stop succeeded\n")
+
+	case <-time.After(100 * time.Millisecond):
```

Comment:
Not super happy with this wait. It's not strictly necessary to include it, but if we don't, then we don't show any opportunity for the soft stop to work at all, so I put it in.